### PR TITLE
Pre-emptively fix synapse.storage.types.Connection for future mypy release

### PR DESCRIPTION
**Pull request by ShadowJonathan**
> Fix the Connection protocol according to typeshed's assertions about sqlite3.Connection

Starting with `mypy` `0.790`, `typeshed` has [updated](https://github.com/python/typeshed/commit/baaffed1ac06d14c72d8e3c621f9721e844a4580) it's assertions about [`sqlite.Connection.__exit__`'s type signature](https://github.com/python/typeshed/blob/baaffed1ac06d14c72d8e3c621f9721e844a4580/stdlib/2and3/sqlite3/dbapi2.pyi#L175), which means that in the future mypy releases, this error will show up.

This is related to #8576, as trying to fix that (by bumping the mypy release) will likely stumble across the issue this PR is fixing.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

**Comment by richvdh** [comment]
:thinking: I thought `__exit__` was supposed to return bool, as per https://docs.python.org/3/library/stdtypes.html#contextmanager.__exit__.

I guess a `None` is a *kind* of falsy value...

**Comment by ShadowJonathan** [comment]
huh, weird that typeshed asserts otherwise then :thinking:

File: synapse/storage/types.py [hunk 1]
```diff
@@ -65,5 +65,5 @@ def rollback(self, *args, **kwargs) -> None:
     def __enter__(self) -> "Connection":
         ...
 
-    def __exit__(self, exc_type, exc_value, traceback) -> bool:
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
```
Comment (richvdh) [inline synapse/storage/types.py:68]: maybe this should be `Optional[bool]` ?

Comment (ShadowJonathan) [reply synapse/storage/types.py:68]: i was *just* thinking about that, maybe it should, yeah

File: synapse/storage/types.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Iterable, Iterator, List, Tuple
+from typing import Any, Iterable, Iterator, List, Tuple, Optional
```
Comment (richvdh) [inline synapse/storage/types.py:15]:
```suggestion
from typing import Any, Iterable, Iterator, List, Optional, Tuple
```

**Review by richvdh — APPROVED** [review]
lgtm assuming CI passes now

**Pull request merged**